packages/keyring/src/keyring.ts

```diff
@@ -0,0 +1,319 @@
+    }
+
+    return newAddressData
+  }
+
+  private _deriveAddressesInGroup = (group: number, amount: number, skipIndexes: number[]): SensitiveAddressData[] => {
+    const addresses = []
+    const skipAddressIndexes = Array.from(skipIndexes)
+
+    for (let j = 0; j < amount; j++) {
+      const newAddress = this._generateAddress({ group, skipAddressIndexes })
+      addresses.push(newAddress)
+      skipAddressIndexes.push(newAddress.index)
+    }
+
+    return addresses
+  }
+
+  private _getNonSensitiveAddressData = ({
+    hash,
+    index,
+    publicKey
+  }: SensitiveAddressData): NonSensitiveAddressData => ({ hash, index, publicKey })
+
+  private _initFromMnemonic = (mnemonic: Uint8Array | null, passphrase: string) => {
+    if (this.root) throw new Error('Keyring: Secret recovery phrase already provided')
+    if (!mnemonic) throw new Error('Keyring: Secret recovery phrase not provided')
+
+    const seed = metamaskBip39.mnemonicToSeedSync(mnemonic, wordlist, passphrase)
+    this.hdWallet = HDKey.fromMasterSeed(seed)
+    this.root = this.hdWallet.derive(this.hdPath)
+
+    passphrase = ''
+  }
+
+  private _deriveAddressAndKeys = (addressIndex: number): SensitiveAddressData => {
+    if (!this.root) throw new Error('Keyring: Cannot derive address and keys, secret recovery phrase is not provided')
+    if (!isAddressIndexValid(addressIndex)) throw new Error('Invalid address index path level')
+
+    const keyPair = this.root.deriveChild(addressIndex)
+
+    if (!keyPair.publicKey) throw new Error('Keyring: Missing public key')
+    if (!keyPair.privateKey) throw new Error('Keyring: Missing private key')
+
+    const publicKey = bytesToHex(keyPair.publicKey)
+    const privateKey = keyPair.privateKey
+    const hash = blake.blake2b(Uint8Array.from(keyPair.publicKey), undefined, 32)
+    const type = new Uint8Array([0])
+    const bytes = new Uint8Array(type.length + hash.length)
+    bytes.set(type, 0)
+    bytes.set(hash, type.length)
+    const address = bs58.encode(bytes)
+
+    return { hash: address, publicKey, privateKey, index: addressIndex }
+  }
+}
+
+export const keyring = new Keyring()
+
+const splitResultsArrayIntoOneArrayPerGroup = (array: boolean[], chunkSize: number): boolean[][] => {
+  const chunks = []
+  let i = 0
+
+  while (i < array.length) {
+    chunks.push(array.slice(i, i + chunkSize))
+    i += chunkSize
+  }
+
+  return chunks
+}
+
+const getGapFromLastActiveAddress = (
+  addresses: SensitiveAddressData[],
+  results: boolean[],
+  startingGap = 0
+): { gap: number; activeAddresses: SensitiveAddressData[] } => {
+  let gap = startingGap
+  const activeAddresses = []
+
+  for (let j = 0; j < addresses.length; j++) {
+    const address = addresses[j]
+    const isActive = results[j]
+
+    if (isActive) {
+      activeAddresses.push(address)
+      gap = 0
+    } else {
+      gap++
+    }
+  }
+
+  return {
+    gap,
+    activeAddresses
+  }
+}
+
+const getActiveAddressesResults = async (
```

Comment:
Maybe it's just me but I find it weird to have a call to an endpoint in the keyring package... 🤔

Reply:
Hm, yeah. This code is a simple adaptation of what we had before in the shared-crypto. I'll look into alternative solutions (maybe passing the keyring as a callback)

Reply:
Addressed in https://github.com/alephium/alephium-frontend/commit/1ce954a1147ac03ee987bfd9c7ac33149c1577c6 in PR https://github.com/alephium/alephium-frontend/pull/504